Source/WTF/wtf/ZippedRange.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/IndexedRange.h>
+
+namespace WTF {
+
+template<typename IteratorA, typename IteratorB> class ZippedRangeIterator {
+public:
+    ZippedRangeIterator(IteratorA&& iteratorA, IteratorB&& iteratorB)
+        : m_iteratorA(WTFMove(iteratorA))
```

Comment:
Shouldn't this be std::forward because that `IteratorA` is a template parameter? If the caller didn't pass an-r-value reference, moving the value out would be confusing / error-prone.

Reply:
Apologies if I have this wrong, because it's new to me, but here's my understanding:

In the context of `ZippedRangeIterator<IteratorA, IteratorB>:: ZippedRangeIterator<IteratorA, IteratorB>()`, `IteratorA` and `IteratorB` are known types, not deduced types. Therefore, `IteratorA&&` and `IteratorB&&` mean rvalue-reference, and not universal reference. Therefore, the caller is required to pass us an rvalue-reference.

(In contrast, if we had a free function `template<typename IteratorA, typename IteratorB> makeZippedRangeIterator(IteratorA&& iteratorA, IteratorB&& iteratorB)`, in the context of that free function `IteratorA` and `IteratorB` would not be known types, and would be deduced types, and `IteratorA&&` and `IteratorB&&` would mean universal reference.)